base/mmap.jl

```diff
@@ -1,172 +1,181 @@
+    flags = shared ? MAP_SHARED : MAP_PRIVATE
+    if s == INVALID_HANDLE_VALUE
+        flags |= MAP_ANONYMOUS
         prot = PROT_READ | PROT_WRITE
+    else
+        mode = ccall(:fcntl,Cint,(Cint,Cint),s,F_GETFL)
+        systemerror("fcntl F_GETFL", mode == -1)
+        mode = mode & 3
+        prot = mode == 0 ? PROT_READ : mode == 1 ? PROT_WRITE : PROT_READ | PROT_WRITE
+        if prot & PROT_READ == 0
+            throw(ArgumentError("mmap requires read permissions on the file (choose r+)"))
+        end
     end
-    if prot & PROT_READ == 0
-        throw(ArgumentError("mmap requires read permissions on the file (choose r+)"))
-    end
-    flags = MAP_SHARED
     return prot, flags, (prot & PROT_WRITE) > 0
 end
 
-# Mmapped-array constructor
-function mmap_array{T,N}(::Type{T}, dims::NTuple{N,Integer}, s::IO, offset::FileOffset; grow::Bool=true)
-    prot, flags, iswrite = mmap_stream_settings(s)
-    len = prod(dims)*sizeof(T)
-    if len > typemax(Int)
-        throw(ArgumentError("file is too large to memory-map on this platform"))
-    end
-    if iswrite && grow
-        pmap, delta = mmap_grow(len, prot, flags, fd(s), offset)
-    else
-        pmap, delta = Libc.mmap(len, prot, flags, fd(s), offset)
+# Before mapping, grow the file to sufficient size
+# Note: a few mappable streams do not support lseek. When Julia
+# supports structures in ccall, switch to fstat.
+grow!(::AnonymousMmap,o::Integer,l::Integer) = return
+function grow!(io::IO, offset::Integer, len::Integer)
+    pos = position(io)
+    filelen = filesize(io)
+    if filelen < offset + len
+        write(io, Base.zeros(UInt8,(offset + len) - filelen))
+        flush(io)
     end
-    A = pointer_to_array(convert(Ptr{T}, UInt(pmap)+delta), dims)
-    finalizer(A,x->Libc.munmap(pmap,len+delta))
-    return A
+    seek(io, pos)
+    return
 end
+end # @unix_only
+
+@windows_only begin
 
+const PAGE_READONLY          = UInt32(0x02)
+const PAGE_READWRITE         = UInt32(0x04)
+const PAGE_WRITECOPY         = UInt32(0x08)
+const PAGE_EXECUTE_READ      = UInt32(0x20)
+const PAGE_EXECUTE_READWRITE = UInt32(0x40)
+const PAGE_EXECUTE_WRITECOPY = UInt32(0x80)
+const FILE_MAP_COPY          = UInt32(0x01)
+const FILE_MAP_WRITE         = UInt32(0x02)
+const FILE_MAP_READ          = UInt32(0x04)
+const FILE_MAP_EXECUTE       = UInt32(0x20)
+
+function gethandle(io::IO)
+    handle = Base._get_osfhandle(RawFD(fd(io))).handle
+    systemerror("could not get handle for file to map: $(Base.FormatMessage())", handle == -1)
+    return Int(handle)
 end
 
-### Windows implementation ###
+settings(sh::AnonymousMmap) = utf16(sh.name), sh.readonly, sh.create
+settings(io::IO) = Ptr{Cwchar_t}(C_NULL), isreadonly(io), true
+end # @windows_only
 
-@windows_only type SharedMemSpec
-    name :: AbstractString
-    readonly :: Bool
-    create :: Bool
+# core impelementation of mmap
+# we define a dummy Array type to achieve the Mmap.Array interface
+type Array{T,N}
```

Comment:
Yes, this needed to be here in order to define `Mmap.Array(::Type{T}, dims)`, since otherwise, it replace the `Base.Array` version. Not sure if there's a cleaner way to handle this.